### PR TITLE
[TIR] Refactor BF16Legalize

### PR DESCRIPTION
**Pull request by tqchen**
This PR refactors BF16Legalize to enable more f32 computations.
We also split the BF16Legalize into two steps.

- BF16ComputeLegalize changes all computation to f32 while keeping
the external BF16 storages.
- BF16StorageLegalize changes all storage to u16.

Now BF16 kernels accept tvm.nd.array that are created as bfloat16 type.

### CODE REVIEW

**Comment by tqchen** [comment]
cc @yangulei @Menooker

**Comment by tqchen** [comment]
cc @vinx13

File: tests/python/frontend/onnx/test_forward.py [hunk 1]
```diff
@@ -99,6 +99,14 @@ def get_tvm_output_with_vm(
             freeze_params=freeze_params,
             convert_config=convert_config,
         )
+        # handle the bfloat16 so we explicitly allocate
+        # bfloat16 arrays as input
+        for i, param in enumerate(mod["main"].params):
+            if param.type_annotation.dtype == "bfloat16":
+                input_data[i] = tvm.nd.empty(input_data[i].shape, "bfloat16").copyfrom(
+                    input_data[i]
+                )
```
Comment (junrushao) [inline tests/python/frontend/onnx/test_forward.py:108]: why are we adding ONNX tests in this PR though?

**Review by junrushao — APPROVED** [review]

File: tests/python/frontend/onnx/test_forward.py (hunk 1 above)
Comment (tqchen) [reply tests/python/frontend/onnx/test_forward.py:108]: This is needed to patch up onnx converting bf16 to uint16

File: src/target/llvm/codegen_llvm.cc [hunk 2]
```diff
@@ -828,6 +828,10 @@ void CodeGenLLVM::CreateSerialFor(llvm::Value* begin, llvm::Value* end, llvm::Va
 llvm::Value* CodeGenLLVM::CreateCast(DataType from, DataType to, llvm::Value* value) {
   llvm::Type* target = DTypeToLLVMType(to);
   if (value->getType() == target) return value;
+  // TODO(tvm-team): consider add native support
+  ICHECK(!from.is_bfloat16()) << "BF16 needs to be storaged lowered first";
+  ICHECK(!to.is_bfloat16()) << "BF16 needs to be storaged lowered first";
+
```
[the remark this answers is not in the document]
Comment (tqchen) [reply src/target/llvm/codegen_llvm.cc:834]: Good point, already left as a todo

**Pull request merged**